File: First Semester/HW7/Task2/Task2/stack.cpp
#include <iostream>
#include "stack.h"

using namespace std;

struct StackElement
{
	ElementType value;
	StackElement *next;
};

struct Stack
{
	StackElement *head;
};

Stack *create()
{
	Stack *stack = new Stack;
	stack->head = nullptr;
	return stack;
}

void push(Stack *stack, ElementType value)
{
	StackElement *newElement = new StackElement;
	newElement->value = value;
	newElement->next = stack->head;
	stack->head = newElement;
}

void pop(Stack *stack)
{
	StackElement *tmp = stack->head;
	stack->head = stack->head->next;
	delete tmp;
}

ElementType top(Stack *stack)
{
	return stack->head->value;
}

bool stackIsEmpty(Stack *stack)
{
	if (stack->head == nullptr)
		return true;
	else
		return false;
}

void deleteStack(Stack *stack)
{
	while (stack->head != nullptr)
		pop(stack);
	delete stack;
}

File: Second Semester/HW2/Stack/Stack/Stack.cs
using System;

namespace HW2_Stack
{
    public class Stack<Type>
    {
        private StackElement head;

        private class StackElement
        {
            /// <summary>
            /// value of stack element 
            /// </summary>
            public Type value { get; set; }
            /// <summary>
            /// link to the next stack element
            /// </summary>
            public StackElement next { get; set; }
        }

        /// <summary>
        /// Function, adding element in stack
        /// </summary>
        /// <param name="element"></param>
        public void Push(Type value)
        {
            StackElement newElement = new StackElement();
            newElement.next = head;
            newElement.value = value;
            head = newElement;
        }

        /// <summary>
        /// Function, deleting head element in stack
        /// </summary>
        public void Pop()
        {
            if (head == null)
            {
                throw new InvalidOperationException("Stack is empty");
            }
            head = head.next;
        }

        /// <summary>
        /// Function, checking emptiness of stack
        /// </summary>
        /// <returns></returns>
        public bool IsEmpty()
        {
            return head == null;
        }

        /// <summary>
        /// Function, returning head element in stack
        /// </summary>
        /// <returns></returns>
        public Type Peek()
        {
            return head.value;
        }

        /// <summary>
        /// Function, printing stack on console
        /// </summary>
        public void PrintStack()
        {
            StackElement tmp = head;
            while (tmp.next != null)
            {
                Console.Write("{0} ", tmp.value);
                tmp = tmp.next;
            }
            Console.WriteLine();
        }
    }
}
